[pruning][docs] Update README.md for structured pruning

jcaip opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #90403

Summary:

I wrote a tutorial of how to use structured pruning flow as part of BE week

Test Plan:

Reviewers:

Subscribers:

Tasks:

Tags:

jcaip committed:
[pruning][docs] Update README.md for structured pruning

Summary:

I wrote a tutorial of how to use structured pruning flow as part of BE week

Test Plan:

Reviewers:

Subscribers:

Tasks:

Tags:

[ghstack-poisoned]

HDCharles commented on torch/ao/pruning/_experimental/pruner/README.md:
```diff
@@ -1,93 +1,263 @@
+
+## Intro / Motivation
+
+**pruning** is the process of removing weights / connections in a neural network in order to make it run faster.
```
I would perhaps elaborate on this terminology, it kind of assumes the user is familiar how we tend to think about NNs, also weights could mean entire tensors, zeroing out values, or changing dimensionality, a bit ambiguous

HDCharles commented on torch/ao/pruning/_experimental/pruner/README.md:
```diff
@@ -1,93 +1,263 @@
+
+There are two types of pruning that we support: **structured pruning** and **unstructured pruing**
+
+* **Unstructured pruning** refers to pruning any arbritary weight
```
would elaborate on this as well, could be interpreted as us removing the weight tensor entirely

HDCharles replied:
also I would direct users to a different resource and emphasize that this guide is about structured pruning

HDCharles commented on torch/ao/pruning/_experimental/pruner/README.md:
```diff
@@ -1,93 +1,263 @@
+There are two types of pruning that we support: **structured pruning** and **unstructured pruing**
+
+* **Unstructured pruning** refers to pruning any arbritary weight
+* **Structured pruning** refers to pruning by removing entire rows / columns of a weight.
```
should be "of a weight tensor" i think it also should be indicated that these changes can propagate onwards

HDCharles commented on torch/ao/pruning/_experimental/pruner/README.md:
```diff
@@ -1,93 +1,263 @@
+
+Structured pruning offers several advantages over unstructured pruning at the cost of lower granuality:
+1. We can store the masks more efficeintly in structured pruning.
+2. Structured pruning relies on removing columns / rows to a speedup whereas unstructured pruning requires custom kernels.
```
could make this clearer to emphasize that structured pruning can use the same kernels as a normal, 

also

"to a speedup" should probably be "to obtain a speedup"

also, significantly less importantly, I think generally "speed-up" is used rather than "speedup"

HDCharles commented on torch/ao/pruning/_experimental/pruner/README.md:
```diff
@@ -1,93 +1,263 @@
+1. We can store the masks more efficeintly in structured pruning.
+2. Structured pruning relies on removing columns / rows to a speedup whereas unstructured pruning requires custom kernels.
+
+Pruning is usually followed by a retraining or fine-tuning step.
```
this line feels a bit out of place, perhaps walk a user through a generic pruning process??

HDCharles commented on torch/ao/pruning/_experimental/pruner/README.md:
```diff
@@ -1,93 +1,263 @@
+
+## What do I need to use Structured Pruning?
+
+**Your model must be FX symbolically traceable**
```
how can a user check this? or get more info about it?

HDCharles commented on torch/ao/pruning/_experimental/pruner/README.md:
```diff
@@ -1,93 +1,263 @@
+If you are looking to prune a different pattern, for example `(conv2d -> activation -> pool -> dropout -> conv2d)`, you can do this by creating modifying the pattern dict that we provide to the pruner, see [here](#writing-custom-patterns-and-pruning-functions-for-structured-pruning)
+
+It is up to the user to define the following two items.
+1. What layers are you looking to prune
```
seems like this may be better covered in the link above about custom pruning, feels a bit out of place, especially without any more details

HDCharles commented on torch/ao/pruning/_experimental/pruner/README.md:
```diff
@@ -1,93 +1,263 @@
+
+If you are looking to prune a different pattern, for example `(conv2d -> activation -> pool -> dropout -> conv2d)`, you can do this by creating modifying the pattern dict that we provide to the pruner, see [here](#writing-custom-patterns-and-pruning-functions-for-structured-pruning)
+
+It is up to the user to define the following two items.
```
this feels a little out of place

HDCharles commented on torch/ao/pruning/_experimental/pruner/README.md:
````diff
@@ -1,93 +1,263 @@
+```
 
-The Base Pruner inherits from the Base Sparsifier.
+Let's walk through this line by line to understand what the different parts are doing.
````
may be better to walk line by line with comments in the code block, as is, this section seems less line-by-line and more just going over the main parts of the APII

HDCharles commented on torch/ao/pruning/_experimental/pruner/README.md:
```diff
@@ -1,93 +1,263 @@
 
-# Motivation
+To specify the layers to prune we just need the fully qualified name (FQN) of the tensor you are looking to prune in the module.
+You can get the FQN of a tensor by printing out `model.named_parameters()`.
```
may want to define what fqn means

HDCharles commented on torch/ao/pruning/_experimental/pruner/README.md:
```diff
@@ -1,93 +1,263 @@
+This `update_mask` function contains the user logic for picking what weights to prune.
 
-![prune example](./images/prune_2.png)
+One common pruning criteria is to use the **saliency** of a row, which is defined as the sum of all the L1 norms of the weights in the row.
```
this is a weird way to describe this, we've used weights to mean weight tensors previously now it looks like this means weight tensor values. Also you can just say L1 norm of the row, L1 norm is an operator on tensors, otherwise I would normally say sum of absolute value of all values in the row

HDCharles commented on torch/ao/pruning/_experimental/pruner/README.md:
```diff
@@ -1,93 +1,263 @@
+The idea is to remove the weights that are small, since they wouldn't contribute much to the final prediction.
 
-The dot product of the masked matrix A (weight) and matrix B (activation) produces the zeros at the sparse locations. However, if we remove the zeros, as in the example shown earlier, the result will change:
+Below we can see an implemented Saliency Pruner
```
I think ideally users will not need to implement their own Pruners, I'm not sure if this is necessary for an introduction

HDCharles commented on torch/ao/pruning/_experimental/pruner/README.md:
```diff
@@ -1,93 +1,263 @@
 
-**PruningParametrization:** After pruning, the shape of the weight changes (some of the output channels are pruned). That means the output of the current layer will have less output layers compared to the original. This means that the next layer should have less input channels.
+Now that we've defined the sparse config for our model, we need to specify the pruning criteria that we will use.
+To do this, we need to extend a `BaseStructuredSparsifier` with a custom `update_mask` function.
```
I think ideally users will not need to implement their own Pruners, I'm not sure if this is necessary for an introduction

HDCharles commented on torch/ao/pruning/_experimental/pruner/README.md:
```diff
@@ -1,93 +1,263 @@
-`step`: applies `update_mask` logic (i.e. prunes the weight matrix)
+For this particular example, we would expect to match the `(nn.Linear, nn.ReLU, nn.Linear)` pattern in our default pattern dict.
 
+Once this is done for all matches, `prune` returns your pruned model:
```
I would maybe clarify that this is not actually what is being returned, it returns a pruned model, the following are useful stats to compare the before/after

HDCharles commented on torch/ao/pruning/_experimental/pruner/README.md:
```diff
@@ -1,93 +1,263 @@
+- conv2d -> pool -> activation -> conv2d
+- conv2d -> adaptive pool -> flatten -> linear
+
+If you are looking to prune a different pattern, for example `(conv2d -> activation -> pool -> dropout -> conv2d)`, you can do this by creating modifying the pattern dict that we provide to the pruner, see [here](#writing-custom-patterns-and-pruning-functions-for-structured-pruning)
```
its a bit unclear to me what 'activation' means in these patterns, i assume it means activation functions but its not defined

HDCharles commented on torch/ao/pruning/_experimental/pruner/README.md:
````diff
@@ -1,93 +1,263 @@
-```
-pruner.squash_mask()
-```
+If you're working with linear/conv2d layers, it's very probable that you just need to add an entry to the pattern dict mapping your pattern to an existing prune_function.
````
unrelated to the documentation, this API seems a bit painful for the average user, if all the user is doing most of the time is indicating that in their (conv-x-conv) pattern that the x op doesn't need to be pruned, the API should probably be able to handle that more elegantly than requiring a user to define a new function that essentially just calls the normal conv2d pruner. If thats not true, i.e. the average user actually needs to carefully handle the x op pruning in most anticipated cases, I think this guide should probably have an example going through that

jcaip replied:
I think the goal is that for most of these patterns, they are already present in the default pruning pattern dict. So for most users this section is unnecessary, and all they have to interact with is the code in the section above.

I think I might split this section and the Implementing a pruner section out, these are more for power users and probably don't need to be in the tutorial.

HDCharles reviewed (approved):
looks good, thanks for working on this!

Commit:
Update on "[pruning][docs] Update README.md for structured pruning"


Summary:

I wrote a tutorial of how to use structured pruning flow as part of BE week

Test Plan:

Reviewers:

Subscribers:

Tasks:

Tags:

[ghstack-poisoned]

jcaip commented:
@pytorchbot merge

jcaip commented:
@pytorchbot merge -f  "documentation changes only"
